src/FormField/MultiLine.php

```diff
@@ -0,0 +1,764 @@
+
+            if (isset($field->ui['form'])) {
+                $type = $field->ui['form'][0];
+            }
+
+            $this->fieldDefs[] = [
+                'field'       => $field->short_name,
+                'type'        => $type,
+                'caption'     => $field->getCaption(),
+                'default'     => $field->default,
+                'isEditable'  => $field->isEditable(),
+                'isHidden'    => $field->isHidden(),
+                'isVisible'   => $field->isVisible(),
+            ];
+        }
+
+        return $m;
+    }
+
+    /**
+     * Returns array of names of fields to automatically include them in form.
+     * This includes all editable or visible fields of the model.
+     *
+     * @param \atk4\data\Model $model
+     *
+     * @return array
+     */
+    protected function getModelFields(\atk4\data\Model $model)
+    {
+        $fields = [];
+        foreach ($model->elements as $f) {
+            if (!$f instanceof \atk4\data\Field) {
+                continue;
+            }
+
+            if ($f->isEditable() || $f->isVisible()) {
+                $fields[] = $f->short_name;
+            }
+        }
+
+        return $fields;
+    }
+
+    public function renderView()
+    {
+        if (!$this->getModel()) {
+            throw new Exception('Multiline field needs to have it\'s model setup.');
+        }
+
+        if ($this->cb->triggered()) {
+            $this->cb->set(function () {
+                try {
+                    return $this->renderCallback();
+                } catch (\atk4\Core\Exception $e) {
+                    $this->app->terminate(json_encode(['success' => false, 'error' => $e->getMessage()]));
+                } catch (\Error $e) {
+                    $this->app->terminate(json_encode(['success' => false, 'error' => $e->getMessage()]));
+                }
+            });
+        }
+
+        $this->multiLine->template->setHTML('Input', $this->getInput());
+        parent::renderView();
+
+        $this->multiLine->vue('atk-multiline',
+                              [
+                                  'data' => [
+                                      'linesField'  => $this->short_name,
+                                      'fields'      => $this->fieldDefs,
+                                      'idField'     => $this->getModel()->id_field,
+                                      'url'         => $this->cb->getJSURL(),
+                                      'eventFields' => $this->eventFields,
+                                      'hasChangeCb' => $this->changeCb ? true : false,
+                                  ],
+                              ]);
+    }
+
+    /**
+     * Render callback.
+     *
+     * @throws ValidationException
+     * @throws \atk4\core\Exception
+     * @throws \atk4\data\Exception
+     */
+    private function renderCallback()
+    {
+        $action = isset($_POST['action']) ? $_POST['action'] : null;
```

Comment:
could this be using callback's trigger value?

Reply:
The callback url is render only once for the component but the same callback url is used for two different action.